language/bytecode-verifier/src/cyclic_dependencies.rs

```diff
@@ -49,17 +43,65 @@ fn verify_module_impl<F: Fn(&ModuleId) -> PartialVMResult<Vec<ModuleId>>>(
         Ok(false)
     }
 
-    let self_id = module.self_id();
     let mut visited_modules = BTreeSet::new();
-    for dep in module.immediate_module_dependencies() {
-        if check_existence_in_dependency_recursive(
-            &self_id,
-            &dep,
-            &immediate_module_dependencies,
+    for item in items_to_explore {
+        if collect_all_with_cycle_detection_recursive(
+            target_module_id,
+            item,
+            immediate_nexts,
             &mut visited_modules,
         )? {
-            return Err(PartialVMError::new(StatusCode::CYCLIC_MODULE_DEPENDENCY));
+            return Err(PartialVMError::new(error_on_cycle));
         }
     }
-    Ok(())
+    Ok(visited_modules)
+}
+
+pub fn verify_module<D, F>(module: &CompiledModule, imm_deps: D, imm_friends: F) -> VMResult<()>
+where
+    D: Fn(&ModuleId) -> PartialVMResult<Vec<ModuleId>>,
```

Comment:
In a follow up, we might want to replace this pair of generics by a trait (e.g., `ModuleDepsResolver`).

This is a pre-existing issue, but it's exacerbated by the addition of another generic to get friends...